### ISSUE TITLE
NVRAM Boot device selection not being saved

### ISSUE THREAD

**Issue by baughmann** [baughmann/designaire-z390-intel-i9-9900k-opencore#4]
Currently, the last selected Boot Device is not being "remembered" in the NVRAM

**Comment by baughmann** [comment]
Even after following the steps in the [Post-Install Guide](https://dortania.github.io/OpenCore-Post-Install/misc/nvram.html) I cannot get this issue resolved

**Comment by baughmann** [comment]
Looks like even grabbing the newest precompiled `SSDT-PMC.aml` doesn't help at all. Not really sure what to do, and I really don't feel like digging into this much. Better to just add a longer boot selector timeout until I get time to diagnose, IMO.

**Comment by schreiberstein** [comment]
Did you try to upgrade to Gigabyte's F9G beta UEFI BIOS?
It supports MSR 0xE2 unlocking - which as far as I understood should allow you to save to NVRAM.
That is, no manual edits are required. It is a regular bios option.

(I do not know how I could test saving to NVRAM. During the macOS installation, it always remembered which pre-booter to choose)

Direct download: https://www.tonymacx86.com/attachments/z390designare-f9g-zip.476399/

Gigabyte has since released a newer beta version - however, it is not recommended from what I have heard.
https://www.tonymacx86.com/threads/success-gigabyte-designare-z390-thunderbolt-3-i7-9700k-amd-rx-580.267551/post-2127054

https://www.tonymacx86.com/threads/success-gigabyte-designare-z390-thunderbolt-3-i7-9700k-amd-rx-580.267551/post-2139806


This is the official file. It had been published on TweakTown forum, but has since been replaced with the more recent beta.
https://www.tweaktownforum.com/forum/tech-support-from-vendors/gigabyte/28656-gigabyte-latest-beta-bios

I hope you are not allergic to tonymacx86. Quite frankly, they have the best resources out there for this particular board.
I wiped my machine two days ago, installed Big Sur Public Beta using your EFI bundle and the F9g beta UEFI BIOS.
The system works flawlessly. A lot better than the Clover-based Mojave setup I had running for 1.5 years using the tools in the tonymacx86 thread. To be fair, opencore was not available at that time.

Cheers and thank you for sharing your EFI bundle,
schreiberstein

**Comment by baughmann** [comment]
@schreiberstein Wowza, thanks for the response!

I've been holding off on the BIOS update purely due to laziness. I'll go ahead and try the `F9h` beta now, cross your fingers!

According to the [OpenCore Post-Install Guide](https://dortania.github.io/OpenCore-Post-Install/misc/nvram.html#verifying-if-you-have-working-nvram) an easy way to test NVRAM is:

```
sudo -s
sudo nvram -c
sudo nvram myvar=test
exit
```
Then, reboot and run
```
nvram -p | grep -i myvar
```

I actually got most of the info I needed for this build (that was not in the OC docs) from the tonymacx86 forum, but just have not kept up with it lately. Honestly, I'm afraid, I'm not really into bleeding edge stuff. I prefer a stable build that will allow me to do my work without looking over my shoulder. I also simply don't have the energy to keep up with all the cool stuff going on in the community these days.

That being said, if you'd like, you can manage an `edge` branch here that uses and recommends all of the latest news. I'm sure a lot of people out there would really appreciate such a branch.

And you're most certainly welcome, thank *you* for the excellent info!

**Comment by schreiberstein** [comment]
I can confirm that NVRAM is working properly. myvar = test after reboot!

I would advice you to go for the F9G (second latest beta) that I mentioned instead of the F9H. My system works fine. Many folks over at tonymacx86 confirmed that it is working fine. I am also not a fan of bleeding edge or unstable systems. There was really no difference / additional configuration from using Catalina. I never used Catalina on this machine. Big Sur, on the other hand, impressed me and I have to say I am quite pleased with it.

**Comment by baughmann** [comment]
Well I finally got back into my system. Took me a minute to realize it was just the fact that the BIOS settings get reset after an update. However, I *still* do not have boot selection remembering. I will keep experimenting...

**Comment by baughmann** [comment]
@schreiberstein Can confirm that this still isn't fixed for me. Have you tried this with a clean EFI, MacOS install, and the latest commit? There must be something separating our systems... maybe a BIOS setting? I'm running on `F9h` successfully now.

**Comment by schreiberstein** [comment]
@baughmann Very strange. I did no modifications to my system. I was glad you prepared this EFI bundle because I do not have the time to explore this myself. And now my system worked instantly. FileVault, iMessage, iCloud etc. That's why it is strange that your's does not...

Core i9 9900K
Z390 Designare
Sapphire Radeon RX 580 Pulse
32GB G.Skill RipJaws 3200 with XMP enabled
Samsung 970 EVO NVME SSD

Look at the tonymacx86 guide and reset your BIOS accordingly.
Load optimized defaults, etc.

I wanted to try out OpenCore but I literally did not want to spend an entire weekend of configuring it.
Hence I just created a USB installer using my MacBook, dropped in your EFI folder, changed the serial, upgraded the BIOS, changed the settings according to the tonymacx86 guide, installed... all is perfectly fine on BigSur. And I suppose Cattle-Lina would work equally well.

Did you set the Other OS setting?

Boot →  Windows 8/10 Features → Other OS

Here are the complete instructions :

https://www.tonymacx86.com/threads/success-gigabyte-designare-z390-thunderbolt-3-i7-9700k-amd-rx-580.267551/

(Look for "BIOS Modifications: Flash Gigabyte BIOS version F4 or Newer" and then F8)

Press F2 to toggle to Advanced Mode.
Press F7 to Load Optimized Defaults.
Now make the following settings:
Tweaker
Advanced CPU Settings → VT-d → Disabled
Extreme Memory Profile (X.M.P.) → Profile 1
Settings
Internal Graphics → Enabled
Above 4G Decoding → Enabled
Thunderbolt(TM) Configuration
Security Level → No Security
This will trigger a 'Save & Exit' dialog so press ESC to close the dialog.
Discrete Thunderbolt(TM) Configuration
Thunderbolt USB Support → Enabled
GPIO3 Force Pwr → Enabled
USB Configuration
XHCI Handoff → Enabled
Boot
Windows 8/10 Features → Other OS
CSM Support → Disabled
Some devices such as GPUs may require CSM Support to be enabled. There is no harm in setting this to either Enabled or Disabled. Use the setting that works best.
Save & Exit → Save Profiles → Save to Profile 1
This preserves the firmware settings for easy recovery.

I even have VT-d enabled. CSM also works fine.

Just reset your BIOS once more to F9G and load defaults. Try to reinstall on a separate drive.

What kind of issues are you experiencing on your osx86 machine?

Cheers,
schreiberstein

**Comment by baughmann** [comment]
Keep in mind, I was also successful in performing that test I recommended to you. However, I was not able to get my boot selection to stay in the NVRAM.

I've applied those BIOS suggestions to no avail. I would guess I probably F'd something up while messing around with trying to get emulated NVRAM working... At some point I'll just have to try a fresh install, probably once the F9 stable release is out.

**Comment by schreiberstein** [comment]
That is unfortunate.
I also do not know what else to try or how to troubleshoot further.

Cheers,
schreiberstein

**Comment by baughmann** [comment]
@schreiberstein Neither do I. I will close this for now and wait for some more BIOS releases, as it looks like Gigabyte is upping their firmware game with some of the early releases they've been coming out with.

**Comment by jfgosselin** [comment]
@baughmann, with the latest Designare BIOS F9i I was able to use CaseySJ EFI out of the box. I had to go to  BIOS Setup --> Boot and disable CFG-Lock.  Also during the first boot I had to  reset the Nvram via the  boot picker (space bar -> Nvram reset).

With  CFG-Lock disabled, you get :

- Native NVRAM support
- Native CPU power management

And my Bluetooth card works now (Broadcom Bcm943602 card):
https://www.amazon.ca/-/fr/gp/product/B07H4Q6864/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1

**Comment by cheneyliuxd** [comment]
Hello, my motherboard, CPU and network card (BT / Wi-Fi Fenvi T919 Bluetooth / Wi-Fi card) are the same as yours. With your EFI file, other functions can work perfectly, but the Bluetooth function cannot be enabled normally. Where is the problem? Because my graphics card is AMD 5500XT, I added the parameter agdpmod=pikera, and nothing else has been changed. Thank you for your advice. My previous functions on catalina are normal.
